### PR TITLE
Transition away from conflicting feature flags in favor of separate crates

### PR DESCRIPTION
**Pull request by neysofu**
Follow-up PR to #416. This PR fixes many of the problems that we've had in the past related to mutually exclusive Cargo features. In particular, it refactors conditional compilation of the `ffi` and `mock` flags into their own separate crates, which allows Cargo to easily avoid conflicts automatically and prepares the ground for fixing #414.

- `svm-sdk` is split into `svm-sdk` and `svm-sdk-mock`. One can easily `extern crate svm_sdk_mock as svm_sdk` to avoid rewriting lots of examples and tests.
- `svm-sdk-host` is split into three crates: `svm-sdk-host` (core trait definition), `svm-sdk-host-ffi` and `svm-sdk-host-mock`.
- Same goes for `svm-sdk-storage`.
- `svm-sdk-macros` is agnostic to both `svm-sdk` and `svm-sdk-mock` with its macros `template` and `template_mock`.

### CODE REVIEW

**Review by YaronWittenstein — COMMENTED** [review]
To my understanding, the new design hurts the testability of new written Templates.
Since on compile-time, you’ll always compile against the real #[template]
or against #[template_mock] one

IMHO using feature flags would be better here since we’d like to write a Template and be able to:
* Compile to Wasm
* Test it using the `svm-sdk-mock`

File: crates/sdk-macros/Cargo.toml [hunk 1]
```diff
@@ -31,23 +31,21 @@ lazy_static = "1.4.0"
...
-mock = ["svm-sdk-host/mock", "svm-sdk-storage/mock"]
 static-alloc = [
     "svm-sdk-types/static-alloc",
     "svm-abi-encoder/static-alloc",
     "svm-abi-decoder/static-alloc",
     "svm-sdk-std/static-alloc",
-    "svm-sdk-host/static-alloc",
-    "svm-sdk-storage/static-alloc",
+    #"svm-sdk-host/static-alloc",
+    #"svm-sdk-storage/static-alloc",
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/sdk-macros/Cargo.toml:42]: Right, missed those. Can be removed.

**Comment by neysofu** [comment]
Good point about source-code reusability both for compiling against FFI and testing purposes (mock). Let met think about a solution.

**Review by YaronWittenstein — APPROVED** [review]

**Review by kobby-pentangeli — APPROVED** [review]
@neysofu Since we're keeping the `Cargo.lock` files (and `.gitignore`-ing the other artifacts), that's the only comment I have on this PR. We can go ahead and merge it.

**Comment by neysofu** [comment]
> @neysofu Since we're keeping the `Cargo.lock` files (and `.gitignore`-ing the other artifacts), that's the only comment I have on this PR. We can go ahead and merge it.

Yes we are for now just to keep moving forward, but I've created an issue and we'll definitely address it in the future.

**Pull request merged**